model/hasfulltablecache.py

```diff
@@ -1,140 +1,91 @@
-
-            # Get the new value of the cache, replacing the value
-            # that turned out to be cls.RESET.
-            cache = getattr(cls, cache_name)
-
-        if cache != cls.RESET:
-            try:
-                obj = cache.get(cache_key)
-            except TypeError as e:
-                # This shouldn't happen. Even if the actual cache was
-                # reset just now, we still have a copy of the 'old'
-                # cache which passed the 'cache != cls.RESET' test.
-                pass
-
-        if not obj:
-            # Either this object didn't exist when the cache was
-            # populated, or the cache was reset while we were trying
-            # to look it up.
-            #
-            # Give up on the cache and go direct to the database,
-            # creating the object if necessary.
-            if lookup_hook:
-                obj, new = lookup_hook()
-            else:
-                obj = None
-            if not obj:
-                # The object doesn't exist and couldn't be created.
-                return obj, new
-
-            # Stick the object in the caches, assuming they're not
-            # currently in a reset state.
-            cls._cache_insert(obj, cls._cache, cls._id_cache)
-
-        if obj and obj not in _db:
-            try:
-                obj = _db.merge(obj, load=False)
-            except Exception as e:
-                logging.error(
-                    "Unable to merge cached object %r into database session",
-                    obj,
-                    exc_info=e,
-                )
-                # Try to look up a fresh copy of the object.
-                obj, new = lookup_hook()
-                if obj and obj in _db:
-                    logging.error("Was able to look up a fresh copy of %r", obj)
-                    return obj, new
-
-                # That didn't work. Re-raise the original exception.
-                logging.error("Unable to look up a fresh copy of %r", obj)
-                raise
-        return obj, new
+        lookup_cache = getattr(cache, cache_name)
+        if cache_key in lookup_cache:
+            cache.stats.hits += 1
+            return lookup_cache[cache_key], False
+        else:
+            cache.stats.misses += 1
+            obj, new = lookup_hook()
+            if obj is not None:
+                cls._cache_insert(obj, cache)
+            return obj, new
+
+    @classmethod
+    def get_cache(cls, _db: Session):
+        """Get cache from database session."""
+        try:
+            cache = _db._palace_cache
+        except AttributeError:
+            _db._palace_cache = cache = {}
+        if cls.__name__ not in cache:
+            cache[cls.__name__] = cls.CacheTuple(
+                {}, {}, SimpleNamespace(hits=0, misses=0)
+            )
+        return cache[cls.__name__]
 
     @classmethod
-    def by_id(cls, _db, id):
+    def by_id(cls, _db: Session, id: int) -> Optional[Base]:
```

Comment:
`_cache_lookup` returns a tuple. I think we need to adjust the type hint here as well.

Reply:
We call `_cache_lookup` in this function then drop, the new value from the tuple, returning only the object.